backend/src/models/User.ts

```diff
@@ -0,0 +1,328 @@
+    saltField: 'salt',
+    activationKeyField: 'activationKey',
+    resetPasswordKeyField: 'resetPasswordKey',
+    incorrectPasswordError: 'Incorrect password',
+    incorrectUsernameError: 'Incorrect username',
+    invalidActivationKeyError: 'Invalid activation key',
+    invalidResetPasswordKeyError: 'Invalid reset password key',
+    missingUsernameError: 'Field %s is not set',
+    missingFieldError: 'Field %s is not set',
+    missingPasswordError: 'Password argument not set!',
+    userExistsError: 'User already exists with %s',
+    activationError: 'Email activation required',
+    noSaltValueStoredError: 'Authentication not possible. No salt value stored in db!'
+};
+const options = defaultAttachOptions;
+
+/**
+ * Initially this table had the fields:
+ * - ID            BIGINT(20)
+ * - ADMIN         TINYINT(1)
+ * - CREATED       DATETIME
+ * - EMAIL         VARCHAR(255)
+ * - LASTMODIFIED  DATETIME
+ * - NAME          VARCHAR(255)
+ * - SLUG          VARCHAR(255)
+ * - CREATOR_ID    BIGINT(20)
+ * - LASTEDITOR_ID BIGINT(20)
+ */
+
+
+
+@Table({
+  timestamps: true
+})
+class User extends Model<User> {
+  @Column({
+    allowNull: false
+  })
+  email: string;
+
+  @Column({
+    allowNull: false
+  })
+  username: string;
+
+  @Column
+  hash: string;
+
+  @Column
+  salt: string;
+
+  @Column
+  activationKey: string;
+
+  @Column
+  resetPasswordKey: string;
+
+  @Column
+  verified: boolean;
+
+  setPassword (password, cb) {
+    if (!password) {
+      return cb(new Error(options.missingPasswordError));
+    }
+
+    const self = this;
+
+    crypto.randomBytes(options.saltlen, function (err, buf) {
+      if (err) {
+        return cb(err);
+      }
+
+      const salt = buf.toString('hex');
+
+      crypto.pbkdf2(password, salt, options.iterations, options.keylen, options.digest, function (err, hashRaw) {
+        if (err) {
+          return cb(err);
+        }
+
+        // TODO: type
+        self.set(options.hashField, new Buffer(hashRaw as any as string, 'binary').toString('hex'));
+        self.set(options.saltField, salt);
+
+        cb(null, self);
+      });
+    });
+  }
+
+  setActivationKey (cb) {
+
+    const self = this;
+
+    if (!options.activationRequired) {
+      return cb(null, self);
+    }
+
+    crypto.randomBytes(options.activationkeylen, function (err, buf) {
+      if (err) {
+        return cb(err);
+      }
+
+      const randomHex = buf.toString('hex');
+      self.set(options.activationKeyField, randomHex);
+      cb(null, self);
+    });
+  }
+
+  authenticate (password, cb) {
+    const self = this;
+
+    // prevent to throw error from crypto.pbkdf2
+    if (!this.get(options.saltField)) {
+      return cb(null, false, { message: options.noSaltValueStoredError });
+    }
+
+    crypto.pbkdf2(password, this.get(options.saltField), options.iterations, options.keylen, options.digest, function (err, hashRaw) {
+      if (err) {
+        return cb(err);
+      }
+
+      // TODO types
```

Comment:
Is now a good time to solve the types issues or do you want to do it at a later point of time?